### PR TITLE
Added exclude_if_none to some app sub-models

### PR DESCRIPTION
**Pull request by orangejenny**
## Technical Summary
The uncompressed CICT app json is about 20% falsy values, largely in case list config since it has a ton of shadow modules. This PR removes many of them. I left boolean attributes alone to avoid confusion, especially because `exclude_if_none` ought to be named `exclude_if_falsy`.

The major goal is to make troubleshooting a bit easier - things like viewing the app in raw doc and doing json diffing. I don't expect this to have much user-facing effect, though it might make wrapping the app a little faster.

### Safety story
This is a fairly mechanical change that should be decently covered by automated tests.

### Automated test coverage

There's coverage of app builds.

### QA Plan

Not requesting QA.

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

**Comment by esoergel** [comment]
Couple test failures - one is that the suite xml was generated differently.  I'd wanna make sure we know how that's possible so we can be confident that couldn't happen in other instances as well (I wonder if something in the suite xml uses `.to_json()`, for instance)

**Comment by orangejenny** [comment]
Updated so this doesn't affect `SchemaProperty` or `SchemaListProperty` attributes.

In [this](https://github.com/dimagi/commcare-hq/blob/8f01b302b1d9cb77db7d908da93cae09b276bcb5/corehq/apps/app_manager/tests/test_suite.py#L203-L213) test code, where an element is being appended to an empty SchemaListProperty attribute, `app.to_json()['sort_elements']` does not contain the new element. If I switch the code to use `detail.sort_elements = [ ... ]` instead of `append`, `app.to_json()['sort_elements']` **does** contain the new element. That's arguably a bug in jsonobject and definitely going to be a source of havoc, so I undid the changes to `SchemaListProperty` attributes.

Also undid the changes to `SchemaProperty` attributes, because those at a minimum are a dict with a `doc_type`, so they're never falsy.

File: corehq/apps/app_manager/models.py [hunk 1]
```diff
@@ -1988,29 +1990,29 @@ class Detail(IndexedSchema, CaseListLookupMixin):
     get_tabs = IndexedSchema.Getter('tabs')
 
     sort_elements = SchemaListProperty(SortElement)
-    filter = StringProperty()
+    filter = StringProperty(exclude_if_none=True)
 
-    instance_name = StringProperty(default='casedb')
+    instance_name = StringProperty(default='casedb', exclude_if_none=True)
```
Comment (snopoke) [inline corehq/apps/app_manager/models.py:1995]: I would expect that properties with defaults would never be None

File: corehq/apps/app_manager/models.py [hunk 2]
```diff
@@ -1014,14 +1014,16 @@ class FormBase(DocumentSchema):
...
     custom_instances = SchemaListProperty(CustomInstance)
     case_references_data = SchemaProperty(CaseReferences)
     is_release_notes_form = BooleanProperty(default=False)
     enable_release_notes = BooleanProperty(default=False)
-    session_endpoint_id = StringProperty()  # See toggles.SESSION_ENDPOINTS
-    function_datum_endpoints = StringListProperty()  # computed datums IDs that are allowed in endpoints
+    session_endpoint_id = StringProperty(exclude_if_none=True)  # See toggles.SESSION_ENDPOINTS
+
+    # computed datums IDs that are allowed in endpoints
+    function_datum_endpoints = StringListProperty(exclude_if_none=True)
```
Comment (snopoke) [inline corehq/apps/app_manager/models.py:1026]: does this not have the same issue as `SchemaListProperty`?

File: corehq/apps/app_manager/models.py [hunk 3]
```diff
@@ -2141,23 +2143,23 @@ class CaseSearch(DocumentSchema):
...
+    search_filter = StringProperty(exclude_if_none=True)
+    search_button_display_condition = StringProperty(exclude_if_none=True)
     default_properties = SchemaListProperty(DefaultCaseSearchProperty)
-    blacklisted_owner_ids_expression = StringProperty()
-    additional_case_types = ListProperty(str)
-    data_registry = StringProperty()
-    data_registry_workflow = StringProperty()           # one of REGISTRY_WORKFLOW_*
-    additional_registry_cases = StringListProperty()  # list of xpath expressions
+    blacklisted_owner_ids_expression = StringProperty(exclude_if_none=True)
+    additional_case_types = ListProperty(str, exclude_if_none=True)
```
Comment (snopoke) [inline corehq/apps/app_manager/models.py:2159]: not directly related: would be nice to use `StringListProperty` for consistency

**Review by snopoke — COMMENTED** [review]
Still some test failures but otherwise looks OK.

File: corehq/apps/app_manager/models.py (hunk 3 above)
Comment (orangejenny) [reply corehq/apps/app_manager/models.py]: updated

File: corehq/apps/app_manager/models.py (hunk 2 above)
Comment (orangejenny) [reply corehq/apps/app_manager/models.py]: Yes, good catch. Fixed.

File: corehq/apps/app_manager/models.py (hunk 1 above)
Comment (orangejenny) [reply corehq/apps/app_manager/models.py]: good point, removed `exclude_if_none` for defaults that aren't falsy

**Comment by snopoke** [comment]
Test failing:
```
FAIL: corehq.apps.app_manager.tests.test_modules:OverwriteModuleDetailTests.test_overwrite_all
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/commcare-hq-ro/corehq/apps/app_manager/tests/test_modules.py", line 185, in test_overwrite_all
    self.assertEqual(self.src_detail.to_json(), dest_detail.to_json())
AssertionError: {'col[773 chars]'def', 'custom_xml': 'ghi'} != {'col[773 chars]'def', 'custom_xml': 'ghi', 'print_template': {'name': 'test'}}
```

**Comment by orangejenny** [comment]
The test failure turns out to be due to [this line](https://github.com/dimagi/commcare-hq/blob/8d1de74ed20de56bc38839007047c628c3bb467b/corehq/apps/app_manager/tests/test_modules.py#L178) assigning a case property directly to `print_template`. It would work if `print_template` was assigned an entire dictionary.

Like the list behavior, this seems risky. Updated PR to only use `exclude_if_none` for string attributes.

**Review by snopoke — APPROVED** [review]

**Pull request merged**